### PR TITLE
Add tooltip and click handler to Funnel Charts

### PR DESCRIPTION
**Pull request by ghsteff**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently Funnel Charts don't support tooltips or an onClick handler to return relevant data

Issue Number: [GOOD-76](https://linear.app/goodcodeus/issue/GOOD-76/add-tooltip-and-click-handler-to-funnel-charts)

## What is the new behavior?
Funnel Charts now support tooltips on hover and an onClick handler to return relevant data

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

https://github.com/reaviz/reaviz/assets/40581813/fc1aa4d7-215b-458f-9ab9-3fbb23507c22

### CODE REVIEW

File: src/common/Tooltip/TooltipArea.tsx [hunk 1]
```diff
@@ -284,7 +290,7 @@ export const TooltipArea = forwardRef<any, Partial<TooltipAreaProps>>(({
       valueScale = yScale;
     }
 
-    const newValue = getClosestPoint(coord, keyScale, transformed, 'x');
+    const newValue = isContinous ? getSelectedSegment(coord, transformed, 'x') : getClosestPoint(coord, keyScale, transformed, 'x');
```
Comment (ghsteff) [inline src/common/Tooltip/TooltipArea.tsx:293]: I wasn't 100% on the name of this prop `isContinous`. What it implies is that the chart has no breaks (aesthetically continous?). It differentiates how this `newValue` for the tooltip is grabbed.

`getClosestPoint` doesn't apply to FunnelCharts like it does to BarCharts. If the closest point is grabbed on a FunnelChart, as soon as you hover more than half way through a segment the tooltip will update to the next segment, so I built a new function `getSelectedSegment`

File: src/common/utils/position.ts [hunk 2]
```diff
@@ -26,6 +26,13 @@ const scaleBandInvert = (scale) => {
   };
 };
 
+export const getSelectedSegment = (position: number, data, positionCoord = 'x') => {
+  // return the item in data that the mouse position is hovering on
+  return data.find((item, i) =>
+    item[positionCoord] <= position && (data[i + 1]?.[positionCoord] > position || data[i + 1] === undefined)
+  );
+};
+
```
Comment (ghsteff) [inline src/common/utils/position.ts:35]: was thinkin about maybe moving this to `utils/selection.ts`? I'll prob keep it here unless anyone has a stronger opinion

Comment (amcdnl) [reply src/common/utils/position.ts:35]: Fine w/ me - Let's add some docs to this function on what exactly it does.

File: src/common/Tooltip/TooltipArea.tsx (hunk 1 above)
Comment (amcdnl) [reply src/common/Tooltip/TooltipArea.tsx:293]: Actually I think the bar chart tooltip is a bug - could we update the bar chart to use this and solve that bug?

File: src/common/Tooltip/TooltipArea.tsx [hunk 3]
```diff
@@ -65,6 +65,11 @@ export interface TooltipAreaProps {
    */
   isRadial?: boolean;
 
+  /**
+   * Whether the area is continous or not.
+   */
+  isContinous?: boolean;
```
Comment (amcdnl) [inline src/common/Tooltip/TooltipArea.tsx:71]: This is not obvious what its doing. Maybe if we need this let's add way more details like why and what its used for.

File: src/FunnelChart/FunnelChart.tsx [hunk 4]
```diff
@@ -22,6 +23,11 @@ export interface FunnelChartProps extends ChartProps {
    * The axis component that renders the funnel axis.
    */
   axis?: ReactElement<FunnelAxisProps, typeof FunnelAxis>;
+
+  /**
+   * Click handler for returning a segment's data.
+   */
+  onClick?: (event: { value: { key: ChartInternalDataTypes, data: ChartShallowDataShape[] }, nativeEvent: MouseEvent }) => void
```
Comment (amcdnl) [inline src/FunnelChart/FunnelChart.tsx:30]: Let's extract the event object so it can be used externally.

**Review by amcdnl — CHANGES_REQUESTED** [review]

File: src/common/Tooltip/TooltipArea.tsx (hunk 1 above)
Comment (ghsteff) [reply src/common/Tooltip/TooltipArea.tsx:293]: Yea I'll try that

File: src/common/Tooltip/TooltipArea.tsx (hunk 3 above)
Comment (ghsteff) [reply src/common/Tooltip/TooltipArea.tsx]: Removed it 👍

File: src/common/Tooltip/TooltipArea.tsx (hunk 1 above)
Comment (ghsteff) [reply src/common/Tooltip/TooltipArea.tsx]: Ok I added support for FunnelCharts to use the getClosestPoint() util and I also fixed that bar chart tooltip bug

File: src/common/utils/position.ts (hunk 2 above)
Comment (ghsteff) [reply src/common/utils/position.ts]: Removed this in favor of adding functionality to getClosestPosition()

**Review by amcdnl — APPROVED** [review]

**Comment by amcdnl** [comment]
@ghsteff - Is this ready for another review?

**Comment by ghsteff** [comment]
> @ghsteff - Is this ready for another review?

No I found a little issue with some of the radial charts that I'm gonna address

**Comment by ghsteff** [comment]
> @ghsteff - Is this ready for another review?

Ok ready for another review

File: src/FunnelChart/FunnelChart.tsx [hunk 5]
```diff
@@ -120,9 +146,10 @@ export const FunnelChart: FC<FunnelChartProps> = ({
       margins={margins}
       containerClassName={containerClassName}
       className={className}
+      onClick={(e: MouseEvent) => handleOnClick(e, getDatas({ chartHeight: height, chartWidth: width }).datas[0])}
```
Comment (amcdnl) [inline src/FunnelChart/FunnelChart.tsx:149]: This is a little complex to be on one line - can we break it out for readability.

File: src/FunnelChart/FunnelChart.tsx [hunk 6]
```diff
@@ -120,9 +146,10 @@ export const FunnelChart: FC<FunnelChartProps> = ({
...
     >
       {renderChart}
-    </ChartContainer>
+    </ChartContainer >
```
Comment (amcdnl) [inline src/FunnelChart/FunnelChart.tsx:152]:
```suggestion
    </ChartContainer>
```

File: src/common/Tooltip/TooltipArea.tsx [hunk 7]
```diff
@@ -137,8 +138,8 @@ export const TooltipArea = forwardRef<any, Partial<TooltipAreaProps>>(({
   placement: placementProp,
   onValueLeave,
   startAngle,
-  endAngle
-}, _) => {
+  endAngle,
```
Comment (amcdnl) [inline src/common/Tooltip/TooltipArea.tsx:141]:
```suggestion
  endAngle
```

File: src/common/types.ts [hunk 8]
```diff
@@ -1,7 +1,15 @@
...
 /**
  * branch between showing something and providing a spec,
  * or not showing it
  */
 export type Visible<T> = { show: true } & T;
 export type Invisible<T> = { show: false } & Partial<T>;
 export type Shown<T> = Visible<T> | Invisible<T>;
+
+export interface ClickEvent { 
+  value: ChartInternalDataShape,
```
Comment (amcdnl) [inline src/common/types.ts:13]:
```suggestion
  value: ChartInternalDataShape;
```

File: src/common/types.ts [hunk 9]
```diff
@@ -1,7 +1,15 @@
...
  * branch between showing something and providing a spec,
  * or not showing it
  */
 export type Visible<T> = { show: true } & T;
 export type Invisible<T> = { show: false } & Partial<T>;
 export type Shown<T> = Visible<T> | Invisible<T>;
+
+export interface ClickEvent { 
+  value: ChartInternalDataShape, 
+  nativeEvent: MouseEvent
```
Comment (amcdnl) [inline src/common/types.ts:14]:
```suggestion
  nativeEvent: MouseEvent;
```

File: src/BarChart/BarSeries/BarSeries.tsx [hunk 10]
```diff
@@ -192,7 +192,7 @@ export const BarSeries: FC<Partial<BarSeriesProps>> = ({
 
   const onMouseMove = useCallback((event) => {
     // Manuallly call mouse move so we don't have to kill bar pointer events
-    ref.current?.onMouseMove(event);
+    ref.current?.triggerMouseMove(event);
```
Comment (amcdnl) [inline src/BarChart/BarSeries/BarSeries.tsx:195]: Was this just a function rename?

Comment (ghsteff) [reply src/BarChart/BarSeries/BarSeries.tsx:195]: There was no function `onMouseMove` within the ref, I added one with the `useImperativeHandle` hook and named it `triggerMouseMove`. It's a wrapper around the `onMouseMove` function within TooltipArea so I gave it a diff name

File: src/FunnelChart/FunnelChart.tsx (hunk 5 above)
Comment (ghsteff) [reply src/FunnelChart/FunnelChart.tsx]: While breaking this out I realized I wasn't using the correct `chartHeight` and `chartWidth` for the Autosize FunnelChart, so I made some changes to store the chart dimensions in state and use them here

**Review by amcdnl — APPROVED** [review]

**Pull request merged**